modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/stat/IgniteStatisticsConfigurationManager.java

```diff
@@ -0,0 +1,518 @@
+
+    /** */
+    private void validate(StatisticsTarget target, GridH2Table tbl) {
+        if (tbl == null) {
+            throw new IgniteSQLException(
+                "Table doesn't exist [schema=" + target.schema() + ", table=" + target.obj() + ']');
+        }
+
+        for (String col : target.columns()) {
+            if (tbl.getColumn(col) == null) {
+                throw new IgniteSQLException(
+                    "Column doesn't exist [schema=" + target.schema() + ", table=" + target.obj() + ", column=" + col + ']');
+            }
+        }
+    }
+
+    /**
+     *
+     */
+    private void updateObjectStatisticConfiguration(StatisticsObjectConfiguration cfg) {
+        try {
+            while (true) {
+                String key = key2String(cfg.key());
+
+                StatisticsObjectConfiguration oldCfg = distrMetaStorage.read(key);
+
+                if (oldCfg != null)
+                    cfg = StatisticsObjectConfiguration.merge(oldCfg, cfg);
+
+                if (distrMetaStorage.compareAndSet(key, oldCfg, cfg))
+                    return;
+            }
+        }
+        catch (IgniteCheckedException ex) {
+            throw new IgniteSQLException("Error on get or update statistic schema", IgniteQueryErrorCode.UNKNOWN, ex);
+        }
+    }
+
+    /** */
+    private void checkLocalStatistics(StatisticsObjectConfiguration cfg, final AffinityTopologyVersion topVer) {
+        try {
+            GridH2Table tbl = schemaMgr.dataTable(cfg.key().schema(), cfg.key().obj());
+
+            if (tbl == null) {
+                // Drop tables handle by onDropTable
+                return;
+            }
+
+            GridCacheContext cctx = tbl.cacheContext();
+
+            AffinityTopologyVersion topVer0 = cctx.affinity().affinityReadyFuture(topVer).get();
+
+            final Set<Integer> parts = cctx.affinity().primaryPartitions(cctx.localNodeId(), topVer0);
+
+            if (log.isDebugEnabled())
+                log.debug("Check local statistics [key=" + cfg.key() + ", parts=" + parts + ']');
+
+            Collection<ObjectPartitionStatisticsImpl> partStats = repo.getLocalPartitionsStatistics(cfg.key());
+
+            Set<String> targetColumns = Arrays.stream(cfg.columns())
+                .map(StatisticsColumnConfiguration::name)
+                .collect(Collectors.toSet());
+
+            Set<Integer> partsToRemove;
+            Set<Integer> partsToCollect;
+
+            if (!F.isEmpty(partStats)) {
+                Set<Integer> storedParts = partStats.stream()
+                    .mapToInt(ObjectPartitionStatisticsImpl::partId)
+                    .boxed()
+                    .collect(Collectors.toSet());
+
+                partsToRemove = new HashSet<>(storedParts);
+                partsToRemove.removeAll(parts);
+
+                partsToCollect = new HashSet<>(parts);
+                partsToCollect.removeAll(storedParts);
+
+                for (ObjectPartitionStatisticsImpl pstat : partStats) {
```

Comment:
1) Naming confusing me (parts is about what we need, but partStats is about what we have, for example).
2) Complicated logic - remove all existing and then add existing with wrong versions. Shall we just remove "what we have" only if it has correct version?